draft-ietf-tls-esni.md

```diff
@@ -375,53 +375,65 @@ This document also defines the "ech_required" alert, which clients MUST send
 
 Some TLS 1.3 extensions can be quite large and having them both in
 ClientHelloInner and ClientHelloOuter will lead to a very large overall size.
 One particularly pathological example is "key_share" with post-quantum
 algorithms. In order to reduce the impact of duplicated extensions, the client
-may use the "outer_extension" extension.
+may use the "outer_extensions" extension.
 
 ~~~
     enum {
-       outer_extension(0xff04), (65535)
+       outer_extensions(0xff04), (65535)
     } ExtensionType;
 
-    struct {
-       ExtensionType outer_extensions<2..254>;
-       uint8 inner_digest<32..255>;
-    } OuterExtensions;
+    ExtensionType OuterExtensions<2..254>;
 ~~~
 
-OuterExtensions MUST only be used in ClientHelloInner. It consists of one or
-more ExtensionType values, each of which reference an extension in
-ClientHelloOuter, and a digest of the complete ClientHelloInner.
+OuterExtensions consists of one or more ExtensionType values, each of which
+reference an extension in ClientHelloOuter.
 
 When sending ClientHello, the client first computes ClientHelloInner, including
-any PSK binders, and then MAY substitute extensions which it knows will be
-duplicated in ClientHelloOuter. To do so, the client computes the digest of the
-entire ClientHelloInner message as:
+any PSK binders. It then computes a new value, the EncodedClientHelloInner, by
+first making a copy of ClientHelloInner. It then clears the legacy_session_id
+field.
+
+The client then MAY substitute extensions which it knows will be duplicated in
+ClientHelloOuter. To do so, the client removes and replaces extensions from
+EncodedClientHelloInner with a single "outer_extensions" extension. Removed
+extensions MUST be ordered consecutively in ClientHelloInner. The list of outer
+extensions, OuterExtensions, includes those which were removed from
+EncodedClientHelloInner, in the order in which they were removed.
+
+Finally, EncodedClientHelloInner is serialized as a ClientHello structure,
+defined in Section 4.1.2 of {{RFC8446}}. Note this does not include the
+four-byte header included in the Handshake structure.
+
+The client-facing server computes ClientHelloInner by reversing this process.
+First it makes a copy of EncodedClientHelloInner and copies the
+legacy_session_id field from ClientHelloOuter. It then looks for an
+"outer_extensions" extension. If found, it replaces the extension with the
+corresponding sequence of extensions in the ClientHelloOuter. If any referenced
+extensions are missing or if "encrypted_client_hello" appears in the list, the
+server MUST abort the connection with an "illegal_parameter" alert.
+
+The "outer_extensions" extension is only used for compressing the
+ClientHelloInner. It MUST NOT be sent in either ClientHelloOuter or
+ClientHelloInner.
 
-~~~
-  inner_digest = Expand(Extract("", inner),
-                        "tls13 ech inner digest", Nh)
-~~~
+## Authenticating the Outer ClientHello {#authenticating-outer-clienthello}
```

Comment:
```suggestion
## Authenticating ClientHelloOuter {#authenticating-outer-clienthello}
```

Reply:
Applied manually since I needed to rebase.